### PR TITLE
Clearer error messages when initilizing the requested ONNX Runtime execution provider fails

### PR DESCRIPTION
**Pull request by fxmarty**
# What does this PR do?

This adds more precise error messages with guidance to the user when the ONNX Runtime configuration is badly installed for some execution provider (notably CUDA, Tensorrt).

Note the fuzzy stuff with tokenizers env variable to avoid the warnings https://github.com/huggingface/transformers/issues/5486 , not sure how to do it better.

I did not add tests as this would require to install / uninstall pip packages to run (or have several docker images, on for each case), not sure if I should? What do you think?

Note that there remains an issue: in case `onnxruntime-gpu` is indeed installed but the dependencies are not met, ONNX Runtime will show `CUDAExecutionProvider` and `TensorrtExecutionProvider` in `get_available_providers()` but initializing the inference session with them will fail anyway, **only with a warning message in C++**. I'm wondering if we should not redirect the warning in Python (tried https://stackoverflow.com/questions/24277488/in-python-how-to-capture-the-stdout-from-a-c-shared-library-to-a-variable , is there better?) to raise an error with a better message.

Related issues https://github.com/huggingface/optimum/issues/511 https://github.com/huggingface/optimum/issues/416

### CODE REVIEW

File: optimum/utils/__init__.py [hunk 1]
```diff
@@ -37,6 +40,22 @@ def is_onnxruntime_available():
     return _onnxruntime_available
 
 
+def is_onnxruntime_strict_available():
```
Comment (michaelbenayoun) [inline optimum/utils/__init__.py:43]: Why "strict"?

File: optimum/utils/__init__.py [hunk 2]
```diff
@@ -37,6 +40,22 @@ def is_onnxruntime_available():
...
+    os.environ["TOKENIZERS_PARALLELISM"] = tokenizers_parallelism
+    return result
+
+
+def is_onnxruntime_gpu_available():
+    tokenizers_parallelism = os.environ.get("TOKENIZERS_PARALLELISM", "")
+    os.environ["TOKENIZERS_PARALLELISM"] = "false"
+    result = subprocess.check_output(["pip list | grep onnxruntime-gpu || true"], shell=True).decode("utf-8") != ""
+    os.environ["TOKENIZERS_PARALLELISM"] = tokenizers_parallelism
+    return result
```
Comment (michaelbenayoun) [inline optimum/utils/__init__.py:56]: I think those functions are quite heavy for just checking that.

Comment (fxmarty) [reply optimum/utils/__init__.py:56]: Agreed, so far I have not find a way directly in Python to know whether `onnxruntime` is installed through `onnxruntime` or `onnxruntime-gpu` package. Will look more into it.

File: optimum/utils/__init__.py (hunk 1 above)
Comment (fxmarty) [reply optimum/utils/__init__.py:43]: It is true if and only if the pypi package `onnxruntime is installed` (not `onnxruntime-gpu`)

File: optimum/utils/__init__.py (hunk 2 above)
Comment (JingyaHuang) [reply optimum/utils/__init__.py:56]: Maybe with something shallower? Like:
```python
>>> import pkg_resources
>>> env = {pkg.key for pkg in pkg_resources.working_set}
>>> "onnxruntime" in env
```
What about also adding the check of `onnxruntime-training` ? I think they are in the same category of problem.

Comment (fxmarty) [reply optimum/utils/__init__.py:56]: Thanks! Unfortunately `"onnxruntime" in env` returns True even if there is only onnxruntime-gpu installed. Let me have a look for onnxruntime-training!

File: optimum/utils/__init__.py (hunk 1 above)
Comment (regisss) [reply optimum/utils/__init__.py:43]: I agree with Michael, "strict" sounds a bit weird but I don't have much better ideas either. Would `is_onnxruntime_cpu_available` make sense? Or can we still run things on the GPU with `onnxruntime` but without `onnxruntime-gpu`?

**Comment by fxmarty** [comment]
I changed much of the code, it should be much cleaner now, no call to `subprocess`, no issue with tokenizer. We just check the `capi` from onnxruntime and check the required libs can indeed be loaded or not.

**Review by regisss — APPROVED** [review]
LGTM!

**Pull request merged**